# 6102TA_02 — Ejercicios sobre colecciones y comprensión

## La Profecía del Gran Diccionario Perdido

**Alumno:** Jaime Martret  
**Lenguaje:** Python  
**Entorno:** Jupyter Notebook en Visual Studio Code

## Objetivo

Analizar diferentes fragmentos de código relacionados con listas, diccionarios y comprensiones de Python.

En cada misión se estudiará:

1. Qué pretendía hacer el programa.
2. Qué hace realmente.
3. Si existe un error sintáctico, de ejecución o lógico.
4. Cómo corregir el código.
5. Qué resultado produce la solución corregida.

In [1]:
print("Entorno preparado correctamente")

Entorno preparado correctamente


# Misión 1 — El Caldero de las Pociones

El maestro alquimista guarda las cantidades de esencia mágica utilizadas para fabricar cada poción.

El objetivo es averiguar qué pociones utilizan más de 30 ingredientes en total.

In [2]:
pociones = [
    {"nombre": "Invisibilidad", "ingredientes": [3, 5, 2]},
    {"nombre": "Vuelo", "ingredientes": [10, 8, 9]},
    {"nombre": "Curación", "ingredientes": [6, 4, 5]},
    {"nombre": "Explosión", "ingredientes": [20, 25, 18]}
]

resultado_original = [
    p["nombre"]
    for p in pociones
    if max(p["ingredientes"]) > 30
]

resultado_original

[]

## Análisis del código original

El código no tiene errores sintácticos, por lo que Python puede ejecutarlo correctamente.

El problema es lógico. La condición utiliza:

```python
max(p["ingredientes"]) > 30

In [6]:
resultado_corregido = [
    p["nombre"]
    for p in pociones
    if sum(p["ingredientes"]) > 30
]

resultado_corregido

['Explosión']

## Resultado final

La poción que utiliza más de 30 ingredientes en total es:

- **Explosión**, con un total de 63 ingredientes.

La corrección consiste en sustituir `max()` por `sum()`, porque el ejercicio pide calcular el total y no el ingrediente individual más alto.

# Misión 2 — El Registro de Dragones

Los dragones registran cuánto oro roban cada semana.

El objetivo es calcular la media de oro robado por cada dragón.

In [7]:
dragones = [
    {"nombre": "Smaug", "oro": [500, 800, 700]},
    {"nombre": "Firulais", "oro": [20, 15, 40]},
    {"nombre": "Pepito", "oro": [150, 180, 200]}
]

medias_originales = {
    d["nombre"]: max(d["oro"]) / len(d["oro"])
    for d in dragones
}

medias_originales

{'Smaug': 266.6666666666667,
 'Firulais': 13.333333333333334,
 'Pepito': 66.66666666666667}

## Análisis del código original

El código no tiene errores sintácticos ni produce una excepción.

El problema es lógico, porque utiliza:

`max(d["oro"]) / len(d["oro"])`

Esto significa que toma únicamente la cantidad más alta de oro robada en una semana y la divide entre el número de semanas.

La media correcta debe calcularse sumando todas las cantidades y dividiendo el total entre el número de valores:

`sum(d["oro"]) / len(d["oro"])`

Este error es especialmente peligroso porque el programa devuelve números válidos y aparentemente razonables, aunque sean incorrectos. Al no producirse ningún error de ejecución, el fallo podría pasar desapercibido.

In [8]:
medias_corregidas = {
    d["nombre"]: sum(d["oro"]) / len(d["oro"])
    for d in dragones
}

medias_corregidas

{'Smaug': 666.6666666666666, 'Firulais': 25.0, 'Pepito': 176.66666666666666}

## Resultado final

Las medias correctas de oro robado son:

- **Smaug:** 666.67 monedas.
- **Firulais:** 25 monedas.
- **Pepito:** 176.67 monedas.

La corrección consiste en sustituir `max()` por `sum()`, porque una media se calcula sumando todos los valores y dividiendo entre la cantidad de datos.

Este error es especialmente peligroso porque el programa no falla: devuelve números válidos, pero incorrectos.

# Misión 3 — Los Fantasmas Contables

Los fantasmas cobran según las horas que asustan.

El objetivo es calcular cuánto debe cobrar cada fantasma multiplicando sus horas por el precio de cada hora.

In [9]:
fantasmas = {
    "Casper": {"horas": 30, "precio": 8},
    "Mocos": {"horas": 50, "precio": 5},
    "Pepa": {"horas": 20, "precio": 12}
}

nominas_originales = {
    nombre: datos["horas"] + datos["precio"]
    for nombre, datos in fantasmas.items()
}

nominas_originales

{'Casper': 38, 'Mocos': 55, 'Pepa': 32}

## Análisis del código original

El código no tiene errores sintácticos y puede ejecutarse correctamente.

El problema es lógico, porque utiliza el operador de suma:

`datos["horas"] + datos["precio"]`

Esto suma las horas trabajadas y el precio por hora, pero una nómina debe calcularse multiplicando ambos valores:

`datos["horas"] * datos["precio"]`

Por eso Casper obtiene 38 monedas en el código original:

`30 + 8 = 38`

Sin embargo, debería cobrar:

`30 × 8 = 240`

In [10]:
nominas_corregidas = {
    nombre: datos["horas"] * datos["precio"]
    for nombre, datos in fantasmas.items()
}

nominas_corregidas

{'Casper': 240, 'Mocos': 250, 'Pepa': 240}

## Resultado final

Las nóminas correctas son:

- **Casper:** 240 monedas.
- **Mocos:** 250 monedas.
- **Pepa:** 240 monedas.

La corrección consiste en sustituir el operador `+` por `*`, porque el sueldo se calcula multiplicando las horas trabajadas por el precio de cada hora.

# Misión 4 — El Bosque de los Unicornios

Los unicornios almacenan sus niveles de felicidad en una lista de listas.

El objetivo es obtener únicamente los niveles de felicidad superiores a 20.

In [11]:
felicidad = [
    [10, 15, 20],
    [25, 40, 35],
    [5, 8, 9]
]

altos_originales = [
    fila
    for fila in felicidad
    for nivel in fila
    if nivel > 20
]

altos_originales

[[25, 40, 35], [25, 40, 35], [25, 40, 35]]

## Análisis del código original

El código no tiene errores sintácticos y puede ejecutarse.

El problema es lógico. La comprensión recorre correctamente cada fila y cada nivel:

`for fila in felicidad`  
`for nivel in fila`

También comprueba correctamente si cada nivel es superior a 20:

`if nivel > 20`

Sin embargo, el valor que se añade a la lista es `fila` en lugar de `nivel`.

Por eso, cada vez que un nivel de la segunda fila cumple la condición, se añade la fila completa:

`[25, 40, 35]`

Como los tres valores de esa fila son superiores a 20, la misma fila aparece tres veces.

In [12]:
altos_corregidos = [
    nivel
    for fila in felicidad
    for nivel in fila
    if nivel > 20
]

altos_corregidos

[25, 40, 35]

## Resultado final

Los niveles de felicidad superiores a 20 son:

- 25
- 40
- 35

La corrección consiste en añadir `nivel` a la lista resultante en lugar de `fila`.

La comprensión correcta recorre cada fila, después cada nivel, y conserva únicamente los valores que cumplen la condición `nivel > 20`.